### PR TITLE
plugins/lightline: migrate to mkNeovimPlugin

### PR DESCRIPTION
**Pull request by khaneliman**
Resolves https://github.com/nix-community/nixvim/issues/2063

### CODE REVIEW

File: plugins/statuslines/lightline.nix [hunk 1]
```diff
@@ -1,117 +1,111 @@
...
-in
-{
-  options = {
-    plugins.lightline = {
-      enable = mkEnableOption "lightline";
+  inherit (lib) types;
+  inherit (lib.nixvim) defaultNullOpts mkNullOrOption;
 
-      package = helpers.mkPluginPackageOption "lightline" pkgs.vimPlugins.lightline-vim;
+  type = types.nullOr (
```
Comment (khaneliman) [inline plugins/statuslines/lightline.nix:11]: Not sure of a good name for this

File: plugins/statuslines/lightline.nix [hunk 2]
```diff
@@ -1,117 +1,206 @@
...
+          lineinfo = "%3l=%-2c";
+          line = "%l";
+          column = "%c";
+          close = "%999X X ";
+          winnr = ''%{winnr()}'';
+        }
+        ''
+          Lightline component definitions.
+          Uses 'statusline' syntax.
+          Consult :h 'statusline' for a list of what's available.
```
Comment (MattSturgeon) [inline plugins/statuslines/lightline.nix:86]:
```suggestion
          Lightline component definitions. Uses 'statusline' syntax.

          Consult `:h 'statusline'` for a list of what's available.
```

File: plugins/statuslines/lightline.nix [hunk 3]
```diff
@@ -1,117 +1,206 @@
...
-          modeMap
-          ;
-      };
-    in
-    mkIf cfg.enable {
-      extraPlugins = [ cfg.package ];
-      globals.lightline = mkIf (configAttrs != { }) configAttrs;
-    };
+  callSetup = false;
+  extraConfig = cfg: { globals.lightline = cfg.settings; };
```
Comment (MattSturgeon) [inline plugins/statuslines/lightline.nix:205]: I think we're moving towards preferring "real" aliases for these kinda things, to preserve wrappers such as `mkDefault` if the user for whatever reason defines config in _both_ `plugins.lightline.settings` _and_ `globals.lightline`...

e.g. https://github.com/nix-community/nixvim/blob/052ee66dbb975960cc151db0b5db6a893503582a/plugins/pluginmanagers/lz-n.nix#L195

and https://github.com/nix-community/nixvim/blob/052ee66dbb975960cc151db0b5db6a893503582a/plugins/utils/firenvim.nix#L92

```suggestion
  extraConfig = cfg: {
    globals.lightline = lib.modules.mkAliasAndWrapDefsWithPriority lib.id options.plugins.lightline.settings;
  };
```

**Review by MattSturgeon — CHANGES_REQUESTED** [review]
Looks good overall

File: plugins/statuslines/lightline.nix (hunk 3 above)
Comment (khaneliman) [reply plugins/statuslines/lightline.nix]: Thanks, I didn't realize that's what this function was. So it basically creates a 2 way binding with some sort of priority?

Comment (MattSturgeon) [reply plugins/statuslines/lightline.nix]: > So it basically creates a 2 way binding with some sort of priority

That's what `mkRenamedOptionModule` and `mkAliasOptionModule` do. This function only makes a one-way alias, but it does it in a way that preserves any priority wrappers used on definitions.

Comment (MattSturgeon) [reply plugins/statuslines/lightline.nix]: When I say "wrapped with a priority", I mean how for example `mkDefault` _wraps_ the config value as:

```nix
{
  _type = "override";
  priority = 1000;
  content = "foo";
}
```

File: plugins/statuslines/lightline.nix [hunk 4]
```diff
@@ -1,117 +1,209 @@
...
-            ''';
+        You can use the name of a function defined elsewhere.
+      '';
+      example = ''
+        plugins.lightline = {
+          enable = true;
+          settings.component_function = {
+            readonly = "LightlineReadonly";
           };
+        };
```
Comment (MattSturgeon) [inline plugins/statuslines/lightline.nix:49]: This isn't an option example; option examples are usually _just_ the value of the option. In this case:
```nix
{
  readonly = "LightlineReadonly";
}
```

File: plugins/statuslines/lightline.nix [hunk 5]
```diff
@@ -1,117 +1,209 @@
...
+
+        extraConfigLua = '''
+          function! LightlineReadonly()
+            return &readonly && &filetype !=# 'help' ? 'RO' : '''
+          endfunction
+        ''';
```
Comment (MattSturgeon) [inline plugins/statuslines/lightline.nix:55]: This looks like viml, not lua

File: plugins/statuslines/lightline.nix (hunk 4 above)
Comment (MattSturgeon) [reply plugins/statuslines/lightline.nix:49]: I guess if the purpose of the example is to demonstrate referencing a function defined elsewhere, perhaps that should be documented in the option's description?

Or at least make an example using `literalMD` or `literalExpression`.

File: plugins/statuslines/lightline.nix (hunk 5 above)
Comment (khaneliman) [reply plugins/statuslines/lightline.nix:55]: Yeah, I was just looking into this. Trying to figure out how to translate their viml to lua.

**Review by MattSturgeon — APPROVED** [review]
LGTM, thanks again!

**Comment by khaneliman** [comment]
@mergifyio queue

**Pull request merged**